wagtail/documents/views/documents.py

```diff
@@ -250,41 +251,44 @@ def edit(request, document_id):
     )
 
 
-@permission_checker.require("delete")
-def delete(request, document_id):
-    Document = get_document_model()
-    doc = get_object_or_404(Document, id=document_id)
-
-    if not permission_policy.user_has_permission_for_instance(
-        request.user, "delete", doc
-    ):
-        raise PermissionDenied
+class DeleteView(generic.DeleteView):
+    model = get_document_model()
+    pk_url_kwarg = "document_id"
+    permission_policy = permission_policy
+    permission_required = "delete"
+    header_icon = "doc-full-inverse"
+    usage_url_name = "wagtaildocs:document_usage"
+    delete_url_name = "wagtaildocs:delete"
+    index_url_name = "wagtaildocs:index"
+    page_title = gettext_lazy("Delete document")
 
-    next_url = get_valid_next_url_from_request(request)
+    def user_has_permission(self, permission):
+        return self.permission_policy.user_has_permission_for_instance(
+            self.request.user, permission, self.object
+        )
 
-    if request.method == "POST":
-        doc.delete()
-        messages.success(
-            request,
-            _("Document '%(document_title)s' deleted.") % {"document_title": doc.title},
+    @property
+    def confirmation_message(self):
+        # This message will only appear in the singular, but we specify a plural
+        # so it can share the translation string with confirm_bulk_delete.html
+        return ngettext(
+            "Are you sure you want to delete this document?",
+            "Are you sure you want to delete these documents?",
+            1,
         )
-        return redirect(next_url) if next_url else redirect("wagtaildocs:index")
 
-    return TemplateResponse(
-        request,
-        "wagtaildocs/documents/confirm_delete.html",
```

Comment:
Do you think we should override `get_templates()` and look for this template before `self.template_name` in case people override this template?

I think it's unlikely people override this template though...

We don't have to do this for images and pages because we are still using their templates instead of the generic one. As for snippets, the work in #10206 should cover template customisation in a more robust way.